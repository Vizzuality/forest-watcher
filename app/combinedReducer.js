import * as reducers from 'redux-modules';
import { combineReducers } from 'redux';

/**
 * Separated so we can test the store state without having to mock loads of dependencies.
 */
export const combinedReducer = combineReducers(reducers);
