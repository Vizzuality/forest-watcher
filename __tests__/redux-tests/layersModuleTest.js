import 'react-native';
import thunk from 'redux-thunk';
import configureStore from 'redux-mock-store';
import { combinedReducer } from 'combinedReducer';
import layerReducer, {
  cacheAreaLayer,
  cacheAreaBasemap,
  cacheLayers,
  downloadAreaById,
  getUserLayers,
  refreshAreaCacheById,
  resetCacheStatus,
  setActiveContextualLayer,
  syncLayers
} from 'redux-modules/layers';

describe('Redux Layers Module', () => {
  // Mock Objects:
  const mockArea = {
    name: 'nameMock',
    id: 'areaIDMock',
    application: 'applicationMock', // used to test that all fields are included in payload
    geostore: { id: 'geostoreIDMock'}
  };

  const mockLayer = {
    url: 'urlMock',
    name: 'nameMock',
    id: 'layerIDMock',
  };

  const mockPendingCache = {
    layerIDMock: {
      areaIDMock: false,
      areaIDMock2: false,
      mockAreaID: 'areaIDMock'
    },
    mockLayerID2: {
      areaIDMock: false,
      areaIDMock2: false,
      areaIDMock3: false,
      mockAreaID: 'areaIDMock'
    }
  };

  it('Initial reducer state', () => {
    expect(layerReducer(undefined, { type: 'NONE' })).toMatchSnapshot({ syncDate: expect.any(Number) });
  });

  describe('Redux Snapshot Thunk Actions', () => {
    let initialStoreState;
    let populatedStoreState;
    let configuredStore;
    let store;

    beforeAll(() => {
      // create store
      initialStoreState = combinedReducer(undefined, { type: 'NONE' });
      populatedStoreState = {
        ...initialStoreState,
        areas: {
          ...initialStoreState.areas,
          data: [mockArea]
        },
        layers: {
          ...initialStoreState.layers,
          data: [mockLayer],
          pendingCache: mockPendingCache
        }
      };
      configuredStore = configureStore([thunk]);
    });

    beforeEach(() => {
      // reset store state and clears actions
      store = configuredStore(initialStoreState);
    });

    // if changing this method, change in other tests too
    function mockDispatchAction(state, action, propertyMatcher, test = true) {
      const newStore = configuredStore({ layers: state });
      newStore.dispatch(action);
      const resolvedActions = newStore.getActions();
      let newState = state;
      // should only be one but the loop is used for future changes and so all tests conform.
      resolvedActions.forEach(resolvedAction => {
        newState = layerReducer(newState, resolvedAction);
      });
      if (test) {
        expect(resolvedActions).toMatchSnapshot();
        expect(newState).toMatchSnapshot(propertyMatcher);
      }
      return newState;
    }

    it('cacheAreaLayer', () => {
      store.dispatch(cacheAreaLayer('areaIDMock', 'layerIDMock'));
      expect(store.getActions()).toMatchSnapshot();

      store = configuredStore(populatedStoreState);

      store.dispatch(cacheAreaLayer('areaIDMock', 'layerIDMock'));
      store.dispatch(cacheAreaLayer('areaIDMock1', 'layerIDMock1'));
      expect(store.getActions()).toMatchSnapshot();
    });

    it('cacheAreaBasemap', () => {
      store.dispatch(cacheAreaBasemap('areaIDMock'));
      expect(store.getActions()).toMatchSnapshot();

      store = configuredStore(populatedStoreState);

      store.dispatch(cacheAreaBasemap('areaIDMock'));
      store.dispatch(cacheAreaBasemap('areaIDMock1'));
      expect(store.getActions()).toMatchSnapshot();
    });

    it('cacheLayers', () => {
      store.dispatch(cacheLayers());
      expect(store.getActions()).toMatchSnapshot();

      store = configuredStore(populatedStoreState);

      store.dispatch(cacheLayers());
      expect(store.getActions()).toMatchSnapshot();
    });

    it('downloadAreaById', () => {
      store.dispatch(downloadAreaById('areaIDMock'));
      expect(store.getActions()).toMatchSnapshot();

      store = configuredStore(populatedStoreState);

      store.dispatch(downloadAreaById('areaIDMock'));
      store.dispatch(downloadAreaById('areaIDMock1'));
      expect(store.getActions()).toMatchSnapshot();
    });

    it('getUserLayers', () => {
      store.dispatch(getUserLayers());
      expect(store.getActions()).toMatchSnapshot();
    });

    it('refreshAreaCacheById', () => {
      store.dispatch(refreshAreaCacheById('areaIDMock'));
      expect(store.getActions()).toMatchSnapshot();

      store = configuredStore(populatedStoreState);

      store.dispatch(refreshAreaCacheById('areaIDMock'));
      store.dispatch(refreshAreaCacheById('areaIDMock1'));
      expect(store.getActions()).toMatchSnapshot();
    });

    it('resetCacheStatus', () => {
      // todo full test with existing area in cache
      store.dispatch(resetCacheStatus('areaIDMock'));
      store.dispatch(resetCacheStatus('areaIDMock1'));
      expect(store.getActions()).toMatchSnapshot();
    });

    it('setActiveContextualLayer', () => {
      store.dispatch(setActiveContextualLayer('layerMock', false));
      store.dispatch(setActiveContextualLayer('layerMock', true));
      expect(store.getActions()).toMatchSnapshot();
    });

    it('setActiveContextualLayer full test', () => {
      const propertyMatcher = { syncDate: expect.any(Number) };
      let newState = layerReducer(undefined, { type: 'NONE' });
      newState = mockDispatchAction(newState, setActiveContextualLayer('layerMock', false), propertyMatcher);
      newState = mockDispatchAction(newState, setActiveContextualLayer('layerMock', true), propertyMatcher);
      mockDispatchAction(newState, setActiveContextualLayer('layerMock', true), propertyMatcher);
    });

    it('syncLayers', () => {
      store.dispatch(syncLayers());
      expect(store.getActions()).toMatchSnapshot();

      store = configuredStore({
        ...initialStoreState,
        layers: {
          ...initialStoreState.layers,
          synced: true
        }
      });

      store.dispatch(syncLayers());
      expect(store.getActions()).toMatchSnapshot();
    });
  });
});
