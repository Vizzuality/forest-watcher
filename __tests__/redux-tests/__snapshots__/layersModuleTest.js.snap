// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Redux Layers Module Initial reducer state 1`] = `
Object {
  "activeLayer": null,
  "cache": Object {},
  "cacheStatus": Object {},
  "data": Array [],
  "layersProgress": Object {},
  "pendingCache": Object {},
  "syncDate": Any<Number>,
  "synced": false,
  "syncing": false,
}
`;

exports[`Redux Layers Module Redux Snapshot Thunk Actions cacheAreaBasemap 1`] = `Array []`;

exports[`Redux Layers Module Redux Snapshot Thunk Actions cacheAreaBasemap 2`] = `
Array [
  Object {
    "payload": Object {
      "area": Object {
        "application": "applicationMock",
        "geostore": Object {
          "id": "geostoreIDMock",
        },
        "id": "areaIDMock",
        "name": "nameMock",
      },
      "layer": Object {
        "id": "basemap",
        "url": "https://a.tile.openstreetmap.org/{z}/{x}/{y}.png",
      },
    },
    "type": "layers/CACHE_LAYER_REQUEST",
  },
]
`;

exports[`Redux Layers Module Redux Snapshot Thunk Actions cacheAreaLayer 1`] = `Array []`;

exports[`Redux Layers Module Redux Snapshot Thunk Actions cacheAreaLayer 2`] = `
Array [
  Object {
    "payload": Object {
      "area": Object {
        "application": "applicationMock",
        "geostore": Object {
          "id": "geostoreIDMock",
        },
        "id": "areaIDMock",
        "name": "nameMock",
      },
      "layer": Object {
        "id": "layerIDMock",
        "name": "nameMock",
        "url": "urlMock",
      },
    },
    "type": "layers/CACHE_LAYER_REQUEST",
  },
]
`;

exports[`Redux Layers Module Redux Snapshot Thunk Actions cacheLayers 1`] = `Array []`;

exports[`Redux Layers Module Redux Snapshot Thunk Actions cacheLayers 2`] = `
Array [
  Object {
    "payload": Object {
      "area": Object {
        "application": "applicationMock",
        "geostore": Object {
          "id": "geostoreIDMock",
        },
        "id": "areaIDMock",
        "name": "nameMock",
      },
      "layer": Object {
        "id": "layerIDMock",
        "name": "nameMock",
        "url": "urlMock",
      },
    },
    "type": "layers/CACHE_LAYER_REQUEST",
  },
]
`;

exports[`Redux Layers Module Redux Snapshot Thunk Actions downloadAreaById 1`] = `Array []`;

exports[`Redux Layers Module Redux Snapshot Thunk Actions downloadAreaById 2`] = `
Array [
  Object {
    "payload": Object {
      "application": "applicationMock",
      "geostore": Object {
        "id": "geostoreIDMock",
      },
      "id": "areaIDMock",
      "name": "nameMock",
    },
    "type": "layers/DOWNLOAD_AREA",
  },
  Object {
    "payload": Object {
      "area": Object {
        "application": "applicationMock",
        "geostore": Object {
          "id": "geostoreIDMock",
        },
        "id": "areaIDMock",
        "name": "nameMock",
      },
      "layer": Object {
        "id": "layerIDMock",
        "name": "nameMock",
        "url": "urlMock",
      },
    },
    "type": "layers/CACHE_LAYER_REQUEST",
  },
]
`;

exports[`Redux Layers Module Redux Snapshot Thunk Actions getUserLayers 1`] = `
Array [
  Object {
    "meta": Object {
      "offline": Object {
        "commit": Object {
          "meta": Object {
            "areas": Array [],
          },
          "type": "layers/GET_LAYERS_COMMIT",
        },
        "effect": Object {
          "url": "<API_URL_MOCK>/contextual-layer/?enabled=true",
        },
        "rollback": Object {
          "type": "layers/GET_LAYERS_ROLLBACK",
        },
      },
    },
    "type": "layers/GET_LAYERS_REQUEST",
  },
]
`;

exports[`Redux Layers Module Redux Snapshot Thunk Actions refreshAreaCacheById 1`] = `
Array [
  Object {
    "payload": "areaIDMock",
    "type": "layer/INVALIDATE_CACHE",
  },
]
`;

exports[`Redux Layers Module Redux Snapshot Thunk Actions refreshAreaCacheById 2`] = `
Array [
  Object {
    "payload": "areaIDMock",
    "type": "layer/INVALIDATE_CACHE",
  },
  Object {
    "payload": Object {
      "application": "applicationMock",
      "geostore": Object {
        "id": "geostoreIDMock",
      },
      "id": "areaIDMock",
      "name": "nameMock",
    },
    "type": "layers/DOWNLOAD_AREA",
  },
  Object {
    "payload": Object {
      "area": Object {
        "application": "applicationMock",
        "geostore": Object {
          "id": "geostoreIDMock",
        },
        "id": "areaIDMock",
        "name": "nameMock",
      },
      "layer": Object {
        "id": "layerIDMock",
        "name": "nameMock",
        "url": "urlMock",
      },
    },
    "type": "layers/CACHE_LAYER_REQUEST",
  },
  Object {
    "payload": "areaIDMock1",
    "type": "layer/INVALIDATE_CACHE",
  },
]
`;

exports[`Redux Layers Module Redux Snapshot Thunk Actions resetCacheStatus 1`] = `
Array [
  Object {
    "payload": Object {
      "areaIDMock": Object {
        "completed": false,
        "error": false,
        "progress": 0,
        "requested": false,
      },
    },
    "type": "layer/SET_CACHE_STATUS",
  },
  Object {
    "payload": Object {
      "areaIDMock1": Object {
        "completed": false,
        "error": false,
        "progress": 0,
        "requested": false,
      },
    },
    "type": "layer/SET_CACHE_STATUS",
  },
]
`;

exports[`Redux Layers Module Redux Snapshot Thunk Actions setActiveContextualLayer 1`] = `
Array [
  Object {
    "payload": null,
    "type": "layers/SET_ACTIVE_LAYER",
  },
  Object {
    "payload": "layerMock",
    "type": "layers/SET_ACTIVE_LAYER",
  },
]
`;

exports[`Redux Layers Module Redux Snapshot Thunk Actions setActiveContextualLayer full test 1`] = `
Array [
  Object {
    "payload": null,
    "type": "layers/SET_ACTIVE_LAYER",
  },
]
`;

exports[`Redux Layers Module Redux Snapshot Thunk Actions setActiveContextualLayer full test 2`] = `
Object {
  "activeLayer": null,
  "cache": Object {},
  "cacheStatus": Object {},
  "data": Array [],
  "layersProgress": Object {},
  "pendingCache": Object {},
  "syncDate": Any<Number>,
  "synced": false,
  "syncing": false,
}
`;

exports[`Redux Layers Module Redux Snapshot Thunk Actions setActiveContextualLayer full test 3`] = `
Array [
  Object {
    "payload": "layerMock",
    "type": "layers/SET_ACTIVE_LAYER",
  },
]
`;

exports[`Redux Layers Module Redux Snapshot Thunk Actions setActiveContextualLayer full test 4`] = `
Object {
  "activeLayer": "layerMock",
  "cache": Object {},
  "cacheStatus": Object {},
  "data": Array [],
  "layersProgress": Object {},
  "pendingCache": Object {},
  "syncDate": Any<Number>,
  "synced": false,
  "syncing": false,
}
`;

exports[`Redux Layers Module Redux Snapshot Thunk Actions setActiveContextualLayer full test 5`] = `
Array [
  Object {
    "payload": null,
    "type": "layers/SET_ACTIVE_LAYER",
  },
]
`;

exports[`Redux Layers Module Redux Snapshot Thunk Actions setActiveContextualLayer full test 6`] = `
Object {
  "activeLayer": null,
  "cache": Object {},
  "cacheStatus": Object {},
  "data": Array [],
  "layersProgress": Object {},
  "pendingCache": Object {},
  "syncDate": Any<Number>,
  "synced": false,
  "syncing": false,
}
`;

exports[`Redux Layers Module Redux Snapshot Thunk Actions syncLayers 1`] = `
Array [
  Object {
    "meta": Object {
      "offline": Object {
        "commit": Object {
          "meta": Object {
            "areas": Array [],
          },
          "type": "layers/GET_LAYERS_COMMIT",
        },
        "effect": Object {
          "url": "<API_URL_MOCK>/contextual-layer/?enabled=true",
        },
        "rollback": Object {
          "type": "layers/GET_LAYERS_ROLLBACK",
        },
      },
    },
    "type": "layers/GET_LAYERS_REQUEST",
  },
]
`;

exports[`Redux Layers Module Redux Snapshot Thunk Actions syncLayers 2`] = `Array []`;
